### PR TITLE
fix: Fix linting, integrate `ruff` and typo check `pre-commit` checks

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

After realizing that our current isort setup is misfunctional and doesn't do anything, I started integrating ruff via pre-commit hooks alongside some other useful tools like spell checkers and file line ending sanitizers.

I tried to keep the change as small as possible and therefore disabled some ruff rules (like the ruff `isort` equivalent, ruff format, line ending cleanup, ...). Those can be enabled at a point in time when no PRs as it could yield quite some merge conflicts.

In theory, the `python_comliance` script could be replaced completely by pre-commit hooks, however, in this PR I didn't want to waste too much time on CI adjustments. I just want to have the pre-commit hooks for my sanity.

Avoiding typos also makes sense I guess :)

### CODE REVIEW

File: .clang-tidy
Comment (robinholzi) [inline .clang-tidy:85]: These changes are all made automatically. I will comment on the things I changed manually here.

File: .pre-commit-config.yaml
Comment (robinholzi) [inline .pre-commit-config.yaml:1]: Manual setup of pre-commit hooks (some still disabled to keep diff small for now)

File: _typos.toml
Comment (robinholzi) [inline _typos.toml:6]: Those two words are recognized as typos. As they occur quite often we whitelist them

File: experiments/criteo_online_dataset/gen_pipelines.py
Comment (robinholzi) [inline experiments/criteo_online_dataset/gen_pipelines.py:122]: I think these were changes by black, as we run it on more directories now.

File: modyn/tests/evaluator/internal/dataset/test_evaluation_dataset.py
Comment (robinholzi) [inline modyn/tests/evaluator/internal/dataset/test_evaluation_dataset.py:2]: I made those annotations manually, we already have equivalent ones for pylint

File: scripts/python_compliance.sh
Comment (robinholzi) [inline scripts/python_compliance.sh:7]: In case someone is still using mamba and not micro mamba, I guess forcing them to upgrade isn't to harmful :)

File: setup.cfg
Comment (robinholzi) [inline setup.cfg:29]: Setting that motivated the PR

Comment (robinholzi) [reply setup.cfg:29]: The underlying issue is not addressed here though, as almost every file would have a diff. Waiting for a time where we wouldn't cause merge conflicts.

File: setup.cfg
Comment (robinholzi) [inline setup.cfg:35]: Hasn't been updated for some time now.

File: benchmark/wildtime_benchmarks/data_generation_huffpost.py
Comment (MaxiBoether) [inline benchmark/wildtime_benchmarks/data_generation_huffpost.py:25]: That should not change

File: black.toml
Comment (MaxiBoether) [inline black.toml:14]: Did you change this on purpose? I remember at some point I needed the local wildcards, but I am not sure whether this is from the pre commit hook or done manually

File: modyn/metadata_processor/processor_strategies/processor_strategy_type.py
Comment (MaxiBoether) [inline modyn/metadata_processor/processor_strategies/processor_strategy_type.py:16]: Why this change?

File: modyn/model_storage/internal/grpc/model_storage_grpc_servicer.py
Comment (MaxiBoether) [inline modyn/model_storage/internal/grpc/model_storage_grpc_servicer.py:3]: Why is this only added for the model storage grpc servicer and not the other servicers? (Some leftover from removing ruff changes from this commit?)

File: .pre-commit-config.yaml
Comment (MaxiBoether) [reply .pre-commit-config.yaml:1]: I am a bit torn on pre commit hooks. The reason we had a script + CI checks for linting is that sometimes, I have had the situation that black's autoformat completely destroyed my file. Typically, I do a commit after doing changes and then another commit (or multiple) for linting, such that there is a clean separation between linting and my stuff. If I understand it correctly, now we would enforce ruff being run before we can commit.  Or is it possible that we get a commit without the autoformatting + then a follow up commit with autolinting applied?

File: .pre-commit-config.yaml
Comment (MaxiBoether) [inline .pre-commit-config.yaml:8]: In what relationship is ruff to the other tools? Which tools will be replaced (black, autopep, mypy etc) and which will stay?

File: scripts/python_compliance.sh
Comment (MaxiBoether) [reply scripts/python_compliance.sh:7]: Then we'd need to update some READMEs as well. I'd actually prefer a brief logic that uses micromamba preferably, but if it does not exist checks for mamba to not break everyone's setup (again after the conda switch :D). We can of course emit a warning/hint to get micromamba.

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thank you! This mostly looks good, but I had some smaller questions and comments here and there, especially on using pre commit hooks and micromamba.

File: .pre-commit-config.yaml
Comment (robinholzi) [reply .pre-commit-config.yaml:8]: I suggest removing isort and flake8 as those tools are almost fully covered. Black can arguably also be removed without too many concerns, but as black is fairly fast, it doesn't hurt too much (they are compatible). Pylint does some things differently, I'd keep it for now. On a greenfield project, I'd rather not use it though.

----

https://docs.astral.sh/ruff/faq/#how-does-ruffs-formatter-compare-to-black

### Is the Ruff linter compatible with Black?

Yes. The Ruff linter is compatible with [Black](https://github.com/psf/black) out-of-the-box, as long as the [line-length](https://docs.astral.sh/ruff/settings/#line-length) setting is consistent between the two.

### How does Ruff's linter compare to Pylint?

At time of writing, Pylint implements ~409 total rules, while Ruff implements over 800, of which at least 209 overlap with the Pylint rule set (see: [#970](https://github.com/astral-sh/ruff/issues/970)).

### How does Ruff's linter compare to Flake8?

Ruff can be used as a drop-in replacement for Flake8 when used (1) without or with a small number of plugins, (2) alongside Black, and (3) on Python 3 code.

Ruff can also replace Black, isort, yesqa, eradicate, and most of the rules implemented in pyupgrade.

### How does Ruff's formatter compare to Black?
The Ruff formatter is designed to be a drop-in replacement for [Black](https://github.com/psf/black).

Specifically, the formatter is intended to emit near-identical output when run over Black-formatted code

### How does Ruff compare to Mypy, or Pyright, or Pyre?
Ruff is a linter, not a type checker.

File: .pre-commit-config.yaml
Comment (robinholzi) [reply .pre-commit-config.yaml:1]: `git commit -n` skips the pre-commit hooks. Is that what you are searching for? I also use it rather often locally when making a temporary snapshot of my code. If you don't like pre-commit hooks you can just not install them them :) We just have to define what the CI does and take that as source of truth.

Comment (MaxiBoether) [reply .pre-commit-config.yaml:1]: Ahh, I see. If we can easily skip the pre-commit hooks, then it sounds great!

File: benchmark/wildtime_benchmarks/data_generation_huffpost.py
Comment (robinholzi) [reply benchmark/wildtime_benchmarks/data_generation_huffpost.py:25]: Oh snap, good catch! Looks like a false positive in the typo checker

File: black.toml
Comment (RobinHolzingerQC) [reply black.toml:14]: I changed it on purpose, but we can surely restore the local wildcards.

File: modyn/metadata_processor/processor_strategies/processor_strategy_type.py
Comment (robinholzi) [reply modyn/metadata_processor/processor_strategies/processor_strategy_type.py:16]: pep8 suggests this: see https://docs.astral.sh/ruff/rules/error-suffix-on-exception-name/

File: modyn/model_storage/internal/grpc/model_storage_grpc_servicer.py
Comment (robinholzi) [reply modyn/model_storage/internal/grpc/model_storage_grpc_servicer.py:3]: Not sure, upon manual inspection I didn't find other functions that don't comply with pep8 naming conventions.

File: black.toml
Comment (robinholzi) [reply black.toml]: I changed it on purpose, but we can surely restore the local wildcards.

File: .pre-commit-config.yaml
Comment (MaxiBoether) [reply .pre-commit-config.yaml:8]: Sounds good. Let's remove them in a follow up PR where we enable ruff.

File: scripts/python_compliance.sh
Comment (MaxiBoether) [reply scripts/python_compliance.sh]: This is still open I think

**Review by MaxiBoether — APPROVED** [review]
Generally LGTM

1) shouldn't we put this into main to keep the diff of the other PR small?

2) i think in the python compliance script we still might want to use mamba in case micromamba is not available. everything else LGTM!

**Comment by robinholzi** [comment]
> Generally LGTM
> 
> 1. shouldn't we put this into main to keep the diff of the other PR small?
> 2. i think in the python compliance script we still might want to use mamba in case micromamba is not available. everything else LGTM!

I merged in #447 so that there wouldn't be any merge conflicts later after merging #447 into main. Switching base branches on GitHub still makes sense to hide the the diff that comes from the other branch. I didn't intend to merge this PR into the other one.

**Pull request merged**
